### PR TITLE
golang: consolidate frame writes...

### PR DESCRIPTION
**Pull request by mmihic**
...to coalesce packets for smaller messages

R: @kriskowal

### CODE REVIEW

File: golang/connection.go [hunk 1]
```diff
@@ -418,22 +418,10 @@ func (c *Connection) withStateRLock(f func() error) error {
...
 		frame := c.framePool.Get()
-		if err := frame.Header.read(fhBuf); err != nil {
-			// TODO(mmihic): Should be a protocol error
-			c.connectionError(err)
-			return
-		}
-
-		if _, err := c.conn.Read(frame.SizedPayload()); err != nil {
+		if err := frame.ReadFrom(c.conn); err != nil {
+			c.framePool.Release(frame)
```
Comment (abhinav) [inline golang/connection.go]: Does the frame not need to be released for the success case when the handlers are called?

Comment (mmihic) [inline golang/connection.go]: No, the frame and its buffer is still needed by the upstream handler.

File: golang/frame.go [hunk 2]
```diff
@@ -57,11 +58,60 @@ type FrameHeader struct {
...
+	f.buffer = make([]byte, payloadCapacity+FrameHeaderSize)
+	f.Payload = f.buffer[FrameHeaderSize:]
+	f.headerBuffer = f.buffer[:FrameHeaderSize]
+	return f
+}
+
+// ReadFrom reads the frame from the given io.Reader
+func (f *Frame) ReadFrom(r io.Reader) error {
+	var rbuf typed.ReadBuffer
+	rbuf.Wrap(f.headerBuffer)
```
Comment (abhinav) [inline golang/frame.go]: What's the reason for not using `NewReadBuffer` here?

Comment (mmihic) [inline golang/frame.go]: unnecessary heap allocation in the middle of a core loop

**Comment by abhinav** [comment]
:shipit:

**Comment by kriskowal** [comment]
Thanks!

**Pull request merged**